examples/end_to_end.py

```diff
@@ -1,39 +1,109 @@
-radon = Radon(image_size, angles)
 
-# forward projection
-sinogram = radon.forward(x)
+class TestNet(nn.Module):
+    def __init__(self, channels=3) -> None:
+        super().__init__()
+        self.sino_model = nn.Sequential(
+            nn.Conv2d(1, channels, 5, padding=2),
+            nn.Conv2d(channels, 1, 5, padding=2)
+        )
+        self.img_model = nn.Conv2d(2, 1, 3, padding=1).to(device)
+        self.det_count = 672
+    
+    def forward_sino(self, sino):
+        print('>>> Forwarding sino net.')
+        return self.sino_model(sino)
+    
+    def forward_img(self, img1, img2):
+        print('>>> Forwarding img net.')
+        img = torch.cat([img1, img2], dim=1)
+        return self.img_model(img)
+    
+    def projection(self, img, angles=None, filter=True):
+        if angles is None:
+            angles = np.linspace(0, np.pi * 2, 360, endpoint=False)
+        
+        volume = Volume2D()
+        volume.set_size(img.shape[-2], img.shape[-1])  # [B, C, H, W]
+        radon = FanBeam(self.det_count, angles, volume=volume)
+        sino = radon.forward(img)
+        if filter:
+            sino = radon.filter_sinogram(sino)
+        return sino
+    
+    def backprojection(self, sino, img_shape, angles=None):
+        if angles is None:
+            angles = np.linspace(0, np.pi * 2, 360, endpoint=False)
 
-# apply sino_model to sinograms
-filtered_sinogram = sino_model(sinogram)
+        volume = Volume2D()
+        volume.set_size(img_shape[-2], img_shape[-1])  # [H, W]
+        radon = FanBeam(self.det_count, angles, volume=volume)
+        img = radon.backward(sino)
+        return img
+    
+    def forward(self, sino, img_shape, angles=None):
+        img = self.backprojection(sino, img_shape, angles=angles)
+        sino_pred = self.forward_sino(sino)
+        img_sino = self.backprojection(sino_pred, img_shape, angles=angles)
+        img_pred = self.forward_img(img, img_sino)
+        return sino_pred, img_sino, img_pred
 
-# backprojection
-backprojected = radon.backprojection(filtered_sinogram)
 
-# apply image_model to backprojected images
-y = image_model(backprojected)
 
-# backward works as usual
-loss = criterion(y, x)
-loss.backward()
\ No newline at end of file
+if __name__ == '__main__':
+    import torch.optim as optim
+    from torch.utils.data import DataLoader
+    
+    device = torch.device('cuda')
+    dataset = TestDataset()
+    loader = DataLoader(dataset, batch_size=2, num_workers=4, pin_memory=True, shuffle=True)
+    net = TestNet().to(device)
+    optimizer = optim.Adam(net.parameters(), lr=1e-4)
+    angles = torch.linspace(0, np.pi * 2, 360, requires_grad=False).float()
+    
+    for i, data in enumerate(loader):
+        print(f'Batch: {i}')
+        img_shape = data.shape[2:]
+        with torch.no_grad():
+            sino = net.projection(data.to(device).detach(), angles=angles).detach()
+            img = net.backprojection(sino, img_shape=img_shape, angles=angles).detach()
```

Comment:
This example fails with 

```
Traceback (most recent call last):
  File "/home/beams4/DCHING/Documents/torch-radon/examples/end_to_end.py", line 105, in <module>
    loss.backward()
  File "/home/beams/DCHING/miniconda3/envs/torchradon/lib/python3.12/site-packages/torch/_tensor.py", line 492, in backward
    torch.autograd.backward(
  File "/home/beams/DCHING/miniconda3/envs/torchradon/lib/python3.12/site-packages/torch/autograd/__init__.py", line 251, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/home/beams/DCHING/miniconda3/envs/torchradon/lib/python3.12/site-packages/torch/autograd/function.py", line 288, in apply
    return user_fn(self, *args)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/beams4/DCHING/Documents/torch-radon/src/python/torch_radon/differentiable_functions.py", line 45, in backward
    grad = cuda_backend.forward(grad_x, angles, ctx.tex_cache, ctx.vol_cfg, ctx.proj_cfg, exec_cfg)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: x must be contiguous
```

So the continuity patch must be merged first.

Reply:
This is fixed at `src/python/torch_radon/differentiable_functions.py` by simply forcing the gradient to be contiguous.

Matteo uses this in his v1 version (https://github.com/matteo-ronchetti/torch-radon/blob/master/torch_radon/differentiable_functions.py#L21), but somehow removes these lines in v2 branch.

Reply:
I'm reluctant to silently force tensors to be contiguous because of the performance cost of copying tensors in memory in order to make them contiguous. In the best case, we can figure out a way for this particular example to avoid this error (It's still unclear to me what specifically is causing the back propagation to be non-contiguous). In the worst case, we should warn the user that the input tensors are being forced to be contiguous.